### PR TITLE
Armstrong Number Detection

### LINKED ISSUE #40: Implement Armstrong Number Checker
## Problem Statement:

Write a function that checks if a given integer is an Armstrong number of the first kind.

An Armstrong number of the first kind is a number that is the sum of its own digits each raised to the power of the number of digits. 

For example: 153 is an Armstrong number because 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153
                      1634 is an Armstrong number because 1^4 + 6^4 + 3^4 + 4^4 = 1 +1296 + 81+ 256 = 1634 

**Examples:**

* **Input:** 153
  **Output:** True

* **Input:** 370
 **Output:** True

* **Input:** 121
  **Output:** False

* **Input:** 100
  **Output:** False

### PR DESCRIPTION
**Pull request by MuqadsaT**
---
name: Armstrong Number Detection Review
about: Adds the `is_armstrong_number` function and unit tests to the project
---

**Assignees:** @MuqadsaT

**Links:** [https://github.com/MIT-Emerging-Talent/ET6-foundations-group-31/issues/40] 
Fixed by https://github.com/MIT-Emerging-Talent/ET6-foundations-group-31/pull/47

This pull request introduces a solution for detecting Armstrong numbers in Python:

* **`is_armstrong_number.py`:**
    - Implements the `is_armstrong_number` function to determine if a given number is an Armstrong number.
    - Includes type hints, docstrings, and error handling.
* **`test_is_armstrong_number.py`:**
    - Contains unit tests for the `is_armstrong_number` function, covering:
        - Standard cases (Armstrong and non-Armstrong numbers)
        - Edge cases (zero, one)
        - Defensive tests (negative input, non-integer input)

**To-do:**

* [x] Review and provide any feedback.
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by PyaeLinn01 — APPROVED** [review]

**Comment by PyaeLinn01** [comment]
Well done.

File: solutions/is_armstrong_number.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
```
Comment (PyaeLinn01) [inline solutions/is_armstrong_number.py:2]: The module docstring is clear and provides a concise summary of its purpose.
Consider expanding the explanation slightly to clarify that the function works for positive integers only.

File: solutions/is_armstrong_number.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+"""
+A module for checking if a given number is an Armstrong number.
+
+Module contents:
+    - is_armstrong_number: Checks if a given number is an Armstrong number.
+
+Created on 2025-01-04
+Author: Muqadsa Tahir
+"""
```
Comment (PyaeLinn01) [inline solutions/is_armstrong_number.py:11]: The function docstring is detailed and provides examples. Consider specifying the expected behavior for edge cases, such as n = 0 or extremely large numbers.

File: solutions/is_armstrong_number.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        True  # 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153
+        >>> is_armstrong_number(370)
+        True
+        >>> is_armstrong_number(121)
+        False
+        >>> is_armstrong_number(100)
+        False
+    """
+    # the number should be an integer greater than 0
+    assert n >= 0, "n must be non-negative"
```
Comment (PyaeLinn01) [inline solutions/is_armstrong_number.py:41]: This assertion ensures n is non-negative. Consider raising a ValueError instead of using assert for better exception handling and clarity.

File: solutions/is_armstrong_number.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        n: The number to check.
+
+    Returns -> bool:
+        True if n is an Armstrong number, False otherwise.
+
+    Raises:
+        AssertionError: If n is negative.
+        AssertionError: If n is not an integer.
+
+    Examples:
```
Comment (PyaeLinn01) [inline solutions/is_armstrong_number.py:30]: Consider mentioning that the function will raise an error for negative or non-integer inputs.

File: solutions/tests/test_is_armstrong_number.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:2]: The module docstring is clear and provides a helpful overview of the test cases. Consider formatting the test categories into a bulleted list for better readability.

File: solutions/tests/test_is_armstrong_number.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    def test_large_numbers_1634(self):
+        """Tests if 1634 is an Armstrong number."""
+        self.assertFalse(is_armstrong_number(1634))
+
+    def test_large_numbers_8208(self):
+        """Tests if 8208 is not an Armstrong number."""
+        self.assertFalse(is_armstrong_number(8208))
+
+    def test_armstrong_number_9474(self):
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:73]: The test is appropriate. Consider adding tests for empty strings or strings with non-numeric characters to ensure robust input validation.

File: solutions/tests/test_is_armstrong_number.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        """Tests if 9474 is an Armstrong number."""
+        self.assertFalse(is_armstrong_number(9474))
+
+    def test_non_armstrong_number_121(self):
+        """Tests if 121 is not an Armstrong number."""
+        self.assertFalse(is_armstrong_number(121))
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:79]: Good handling of negative inputs. However, since the implementation uses assert, consider updating the test to expect a ValueError instead of an AssertionError for better consistency.

File: solutions/tests/test_is_armstrong_number.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    - Tests with negative input values.
+    - Tests with non-integer input values (e.g., floats, strings).
+
+Created on 2025-01-04
+Author: Muqadsa Tahir
+"""
+
+import unittest
+
+from ..is_armstrong_number import is_armstrong_number
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:26]: The tests for known Armstrong numbers are thorough. Consider adding tests for larger Armstrong numbers such as 9474, 54748, or 92727 to ensure the function handles them correctly.

File: solutions/is_armstrong_number.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    - is_armstrong_number: Checks if a given number is an Armstrong number.
+
+Created on 2025-01-04
+Author: Muqadsa Tahir
+"""
+
+
+def is_armstrong_number(n: int) -> bool:
+    """
+    Checks if a given number is an Armstrong number.
```
Comment (PyaeLinn01) [inline solutions/is_armstrong_number.py:16]: The docstring provides a detailed description of the function.
Some improvements:
Use ValueError or TypeError explicitly in the "Raises" section instead of AssertionError.
Update the explanation to clarify that it handles numbers with any number of digits, not just three digits.

File: solutions/tests/test_is_armstrong_number.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+    def test_armstrong_number_371(self):
+        """Tests if 371 is an Armstrong number."""
+        self.assertTrue(is_armstrong_number(371))
+
+    def test_armstrong_number_407(self):
+        """Tests if 407 is an Armstrong number."""
+        self.assertTrue(is_armstrong_number(407))
+
+    def test_large_numbers_1634(self):
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:65]: This test seems to misrepresent 1634, which is an Armstrong number (1⁴ + 6⁴ + 3⁴ + 4⁴ = 1634).
Suggested Fix:
def test_large_numbers_1634(self):
    """Tests if 1634 is an Armstrong number."""
    self.assertTrue(is_armstrong_number(1634))

File: solutions/tests/test_is_armstrong_number.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    def test_edge_case_1000(self):
+        """Tests if a 3-digit large n is not an Armstrong number."""
+        self.assertFalse(is_armstrong_number(1000))
+
+    def test_non_integer_input_string(self):
+        """Tests if TypeError is raised for string input."""
+        with self.assertRaises(TypeError):
+            is_armstrong_number("123")
+
+    def test_negative_input(self):
```
Comment (PyaeLinn01) [inline solutions/tests/test_is_armstrong_number.py:106]: As per the updated implementation of is_armstrong_number, negative numbers should raise a ValueError instead of an AssertionError.

Suggested fix:

def test_negative_input(self):
    """Tests if ValueError is raised for negative input."""
    with self.assertRaises(ValueError):
        is_armstrong_number(-1)
